generator/common/beatgen/setup/creator.go

```diff
@@ -0,0 +1,87 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package setup
+
+import (
+	"fmt"
+	"go/build"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	devtools "github.com/elastic/beats/dev-tools/mage"
+	"github.com/pkg/errors"
+)
+
+// CfgPrefix specifies the env variable prefix used to configure the beat
+var CfgPrefix = "NEWBEAT"
+
+// GenNewBeat generates a new custom beat
+// We assume our config object is populated and valid here
+func GenNewBeat(config map[string]string) error {
+	if config["type"] != "beat" && config["type"] != "metricbeat" {
+		return fmt.Errorf("%s is not a valid custom beat type. Valid types are 'beat' and 'metricbeat'", config["type"])
+	}
+
+	genPath := devtools.OSSBeatDir("generator", config["type"], "{beat}")
+	err := filepath.Walk(genPath, func(path string, info os.FileInfo, err error) error {
+		newBase := filepath.Join(build.Default.GOPATH, "src", config["beat_path"])
+		replacePath := strings.Replace(path, genPath, newBase, -1)
+
+		writePath := strings.Replace(replacePath, "{beat}", config["project_name"], -1)
+		writePath = strings.Replace(writePath, ".go.tmpl", ".go", -1)
+		if info.IsDir() {
+			err := os.MkdirAll(writePath, 0755)
+			if err != nil {
+				return errors.Wrap(err, "error creating directory")
+			}
+		} else {
+
+			//dump original source file
+			tmplFile, err := ioutil.ReadFile(path)
```

Comment:
Here there are a bunch of different errors that can be returned. From the ones of the `os.Open` method or the ones of `ioutil.readAll` function. I'd love to ensure that the returned error is descriptive enough so the user is able to solve the problem alone, without looking at docs.

All in all, when generating a beat, an "open file" error might not be descriptive enough about the problem